Implement setseed function

geokollias opened the pull request:
From what I understand, all std's predefined random number generators generate unsigned integer values and their seed methods also expect values of unsigned integral type. In some local experiments with the current PR I see that e.g. all seed values between 0.1-0.9 generate the same sequence of "random" numbers, making the random() function a bit useless if used with that range of inputs. Instead, it's working fine if we provide different unsigned integers as seeds.

I'm mentioning the above because Postgres' setseed function supports only seed values between -1.0 and 1.0, inclusive. I'd like your input about how DuckDB should handle random/setseed:

- Leave it as it is in this PR and expect integers as seeds
- Follow Postgres (-1.0 <= seed value <= 1.0) + use a std predefined random generator (e.g. the one in this PR) + manipulate the seed value so that we end-up with a large int. For example, we could multiply the seed with a large number so that seeds 0.123 & 0.124 become 123000 & 124000 respectively and generate different sequences (in the current PR 0.123 & 0.124 produce the same sequences).
- Follow Postgres (-1.0 <= seed value <= 1.0) + use whatever random engine is used by Postgres (erand48 from stdlib?)

Something else I'm not sure how to handle is the return type of setseed. In Postgres, its return type is void. I'm not sure how to achieve that in DuckDB. Whatever I tried was failing with conversion errors. For the time being, updated it to just return the seed value itself. Any thoughts on how to implement that in DuckDB or if we even care about it?

geokollias committed:
Implement setseed function

Mytherin commented:
> Follow Postgres (-1.0 <= seed value <= 1.0) + use a std predefined random generator (e.g. the one in this PR) + manipulate the seed value so that we end-up with a large int. For example, we could multiply the seed with a large number so that seeds 0.123 & 0.124 become 123000 & 124000 respectively and generate different sequences (in the current PR 0.123 & 0.124 produce the same sequences).

I would use this approach, and indeed just multiply to change the range. `-1.0=0`, `1.0=2^32`. An approach to do this would be something along the lines of `(seed + 1.0) * std::numeric_limits<uint32_t>::max()/2` (after checking that seed is within the proper bounds, of course).

> Something else I'm not sure how to handle is the return type of setseed. In Postgres, its return type is void. I'm not sure how to achieve that in DuckDB. Whatever I tried was failing with conversion errors. For the time being, updated it to just return the seed value itself. Any thoughts on how to implement that in DuckDB or if we even care about it?

It could return SQLNULL, and then return a constant NULL value.

geokollias committed:
Update setseed

Updates:
- Constrain seed value between -1.0 and 1.0, inclusive
- Set return type to SQLNULL

geokollias commented:
Thanks @Mytherin. I followed your suggestions. Please have another look.

Mytherin commented:
Thanks for the changes, it looks good. One more thing: what happens if you input a column into the setseed function that could potentially have multiple values, e.g.:

```SELECT setseed(a) FROM table```


I suppose it should just repeatedly call `setseed` for each value in the table and the resulting seed should be the last value of the table. Could you create a testcase including that as well? After that I think it can be merged. Thanks!

arjenpdevries commented:
> `SELECT setseed(a) FROM table`
> 
> I suppose it should just repeatedly call `setseed` for each value in the table and the resulting seed should be the last value of the table. Could you create a testcase including that as well? After that I think it can be merged. Thanks!

Assuming `setseed` has no other side-effects than changing the seed, would you really want to call it many times? It makes equal sense to just apply it to any value from the table (could even be a random one, but first or last would make intuitive sense even if relations have no order :-)).

Mytherin commented:
> Assuming setseed has no other side-effects than changing the seed, would you really want to call it many times? It makes equal sense to just apply it to any value from the table (could even be a random one, but first or last would make intuitive sense even if relations have no order :-)).

Indeed it is not required, but as you said since calling `setseed` with a large column as input does nothing except set the seed value to the latest one there is also no point for the user to call it with a large column as input, hence I think optimizing this use case is not necessary. Especially since calling it only with the last value is not very trivial in a chunk-wise streaming engine, which I would argue is the expected behavior. 

Note also that the `setseed` function still needs to return a single value per row in the input as well, i.e. if the function is called as follows:

```sql
SELECT setseed(a), a FROM table
```

The setseed function needs to return a value for each row in the table as not to violate relational semantics.

geokollias committed:
Update setseed

geokollias commented:
@Mytherin Pushed another update to address your feedback.

Still not sure the previous behavior was wrong. In case of multiple inputs, `setseed` would pick only the first value (similar to what @arjenpdevries suggested) and would still return a single value (NULL) per input row.

Mytherin commented:
Thanks, it looks good now. Apologies for the delay. 

> Still not sure the previous behavior was wrong. In case of multiple inputs, setseed would pick only the first value (similar to what @arjenpdevries suggested) and would still return a single value (NULL) per input row.

It would pick only the first value of each vector in the input. I would not say it is wrong but it might be unexpected compared to Postgres' implementation.
